rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,350 @@
+
+`JobState`: Describes the current state of the job. The possible values are
+`"running"`, `"pending"`, and `"finished"`. These are the minimal number of
+states needed to describe a job's behavior in respect to this framework. Each
+job type MAY have additional, type specific states, such as `"failed`",
+`"error"`, `"retrying"`, etc.
+
+`Sequence`: a 13 byte value formed by combining the current `Incarnation` of
+the database and the `Versionstamp` of the transaction. Sequences are
+monotonically increasing even when a database is relocated across FoundationDB
+clusters. See (RFC002) for a full explanation.
+
+---
+
+# Framework API
+
+This section describes the job creation and worker implementation APIs. It doesn't
+describe how the framework is implemented. The intended audience is CouchDB
+developers using this framework to implement background jobs for indexing,
+replication, and couch-peruser.
+
+Both the job creation and the worker implementation APIs use a `JobOpts` map to
+represent a job. It MAY also contain these top level fields:
+
+  * `"priority"` : The value of this field will contain the `Priority` value of
+    the job. `Priority` is job-type specific.
+  * `"data"`: An opaque object (map), from the framework's point of view,
+    containing job-type specific data. It MAY contain an update sequence, or an
+    error message, for example.
+  * `"cancel"` : Boolean field defaulting to `false`. If `true` indicates the
+    user intends to stop a job's execution.
+  * `"resubmit"` : Boolean field defaulting to `false`. If `true` indicates
+    the job should be re-submitted.
+
+### Job Creation API ###
+
+```
+add(Type, JobId, JobOpts) -> ok | {error, Error}
+```
+ - Add a job to be executed by a background worker.
+
+```
+remove(Type, JobId) -> ok | not_found
+```
+ - Remove a job. If it is running, it will be stopped, then it will be removed
+   from the job table.
+
+```
+resubmit(Type, JobId) -> ok | not_found
+```
+ - Indicates that the job should be re-submitted for execution.
+
+```
+get_job(Type, JobId) -> {ok, JobOpts, JobState}
+```
+ - Return `JobOpts` and the `JobState`. `JobState` value MAY be:
+  * `"pending"` : This job is pending.
+  * `"running"` : This job is currently running.
+  * `"finished"` : This job has finished running and is not pending.
+
+### Worker Implementation API
+
+This API is to be used when implementing workers for various job types. The general pattern
+is to call `accept()` from something like a job manager, then for each accepted
+job spawn a worker to execute it, and then resume calling `accept()` to get
+other jobs. When a job is running, the worker MUST periodically call `update()`
+to prevent the activity monitor from re-enqueueing it. When the worker decides to stop
+running a job, they MUST call `finish()` to indicate that the job has finished running.
+
+```
+accept(Type[, MaxPriority]) -> {ok, JobId, WorkerLockId} | not_found
+```
```

Comment:
> Its not entirely clear to me how many simultaneous calls to `accept` should be made. There's a mention of a `job manager` but is that required or just a hint at a specific implementation? Is it possible to just spawn `N` consumers that all call `accept` independently?
> 
> Also does `accept` block? When is `not_found` returned? What should happen if `not_found` is returned? Do we sleep for a bit and try again? Is that an unrecoverable error like "I have no idea what `Type` you're talking about." sort of response?

Could spawn any number of consumers each calling `accept`. If any one can dequeue a job they'll return with a lock id, if they can't, they return `not_found`.  They maybe try again after some time. The job manager could be a gen_server that periodically calls accept, and maybe the more jobs it has already dequeued, the longer it sleeps in between accepts.

As the for type not being found, it's possible but it's also not a terminal condition in that it could be that there aren't any jobs of that type added yet. 

> Also, same for `get_job` it seems like it would be better to return `{ok, Job}` here to make client code more ergonomic so we're only carrying around one variable instead of the possible 5 that `finish` takes as arguments.

The `Job` object doesn't have the `JobId` in it. We could return the `Job` here but we definitely need the `JobId`. And since somewhere else we discussed maybe just letting users update the `data` field only, we might have less incentive to add the `Job` to the response.

Reply:
The latest version consolidates those fields into the `Job` opaque object, with functions to set `JobData` and resubmit flag.